Standardizes feather icon

rfbezerra opened the pull request:
This is a effort to resolve #922 
My attempt is to use the library [Font Custom](https://github.com/FontCustom/fontcustom) to solve the manual tasks to update the font icon set.

Using that library, it is possible to automatize the process.
The downside is the need to use a external library.

hampustagerud commented on scripts/feather.sh:
```diff
@@ -1,9 +1,19 @@
 #!/bin/bash -e
 
-node bin/generate-icon Feather/style.css\
+fontcustom compile node_modules/feather-icons/dist/icons\
```
Add something to let users know how to fix this if it fails here, just check the status code and print a link to BUILDING_FEATHER.md if it failed or something 🙂

hampustagerud commented on package.json:
```diff
@@ -72,6 +72,7 @@
     "@babel/cli": "^7.1.2",
     "@babel/core": "^7.1.2",
     "@mdi/font": "^2.8.94",
+    "feather-icons": "^4.10.0",
```
Was this manually inserted? Everything else in ordered by name but this and it stands out a bit, I think yarn and npm inserts everything ordered 🙂

oblador commented:
@rfbezerra Any progress on this? Would be great to get merged :-)

rfbezerra commented:
Hi, sorry for delay
I'm working on this issue, and will have a patch soon

rfbezerra commented:
Well, I made a simple shell script to verify things and warn if anything goes wrong.
It's simple, and any contribution are wellcome.

oblador commented:
@rfbezerra I fixed conflicts and bumped version of Feather Icons. However when I run the IconExplorer with the updated font it looks like the line height is very off. See attached screenshot: 

![Simulator Screen Shot - iPhone 6 - 2019-03-11 at 20 11 10](https://user-images.githubusercontent.com/378279/54150960-04277f00-443a-11e9-89cc-cc716bf04a91.png)

```
$ fontcustom -v
fontcustom-2.0.0
$ fontforge -v
Copyright (c) 2000-2014 by George Williams. See AUTHORS for Contributors.
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
 with many parts BSD <http://fontforge.org/license.html>. Please read LICENSE.
 Based on sources from 23:31 UTC 12-Jan-2019-D.
 Based on source from git with hash: 
fontforge 23:31 UTC 12-Jan-2019
libfontforge 20190112
```

rfbezerra commented:
Hum, I'll look into the issues and return with updates

oblador commented:
Any luck? :-)

rfbezerra commented:
Not yet
I'm trying to talk with owners of tools utilized and trying another solutions.
I guess that error (line height) comes from the TTF font specifications (ascent, descent, etc....), with I have no knowledge about.

rfbezerra commented:
Well, one interest thing.
I found an icon **tablet.svg** with causes weird line height. I don't know why, but if I remove this icon (through generator script) all goes fine.

rfbezerra commented:
Like I mentioned before, one more step is necessary before ttf generation, and consist in transform every svg icon to path. I found an option in Inkscape (Edit -> Select all | Path -> Stroke to path) to do that.
Its a terrible solution because for every svg icon, is necessary to open a Inkscape GUI to to automatize things and then close, so, for this icon set, open and close Inkscape 276 times before ttf generation.
It is possible through command line with a simple script, but you will see Inkscape opening and closing 276 times.
So, I have no idea what to do. Maybe create a separated git repo to handle all and export only things needed to vector-icons to work.
What do you think about?

hampustagerud commented:
@rfbezerra Let's get this merged!

Is there not a CLI tool available to do this conversion? Opening Inkscape is not really optimal but if that is the only way then that's the way it is. I tried googling a little and [elrumordelaluz/outline-stroke-cli](https://www.github.com/elrumordelaluz/outline-stroke-cli) came up, is that a solution? It's an npm package which is easy to include as a `devDependency` 🙂

rfbezerra commented:
I forked the feather repo and did the font generation.
I will mantain in sync with official repo.
https://github.com/rfbezerra/feather

I will update this PR to use this repo

oblador commented:
@rfbezerra I generally don't think it's ideal to use your fork as we would add a dependency on you maintaining it (which you want to do now, but perhaps not in the future). Don't you think it would be possible to do apply the fixes you do in the fork in a script hosted in this repo instead? The updates are currently ran manually anyways so it doesn't matter much if it takes some time to run it. 

Thanks for your perseverance on this issue!

rfbezerra commented:
I did some changes and accidentally found a better solution, using Xvfb.
With command _xvfb-run_ present, inkscape windows goes to virtual frame buffer, not showing in actual display. :+1: 
If not installed Xvfb, all windows will popup.
Actual solution uses GNU Parallel to speedup things and Inkscape app.

Take a look and let me know what you think about.

hampustagerud commented:
I managed to get this to build in about 160 seconds which is quite good but indeed, every icon required inkscape to create a new window which probably added some time!

Dare I assume you are developing this on Linux? 🙂 `xvfb` seems bundled on macOS (or maybe came with XQuartz 🤔) but `xvfb-run` is apparently a Linux specific utility (according to a Google search). I don't know if the command could be converted to a "pure" `xvfb`?

I get an error on every icon:

```
WARNING: ignoring verb EditSelectAll - GUI required for this verb.

** (inkscape-bin:84469): CRITICAL **: static void Inkscape::EditVerb::perform(SPAction *, void *): assertion 'ensure_desktop_valid(action)' failed
WARNING: ignoring verb StrokeToPath - GUI required for this verb.

** (inkscape-bin:84469): CRITICAL **: static void Inkscape::SelectionVerb::perform(SPAction *, void *): assertion 'ensure_desktop_valid(action)' failed
WARNING: ignoring verb FileSave - GUI required for this verb.

** (inkscape-bin:84469): CRITICAL **: static void Inkscape::FileVerb::perform(SPAction *, void *): assertion 'ensure_desktop_valid(action)' failed

Emergency save activated!
Emergency save completed. Inkscape will close now.
```

This seems like a problem related to inkscape cli, nothing you encountered? Adding `--with-gui` to the inkscape command reduces it to:
```
(inkscape-bin:36980): Gtk-WARNING **: cannot open display:
```

This is an X11 problem and X11 is a hassle on macOS which is why I shy away from Inkscape whenever I can. I managed to run `xvfb` but then it generated a different font due to `RANDR extension missing` so I guess we're stuck generating the font with GUI on macOS which is fine. I solved it like this (requires XQuartz to be installed as well, could be added to the README):
```diff
   TEMP_DIR=`mktemp -d`
   echo "Using ${TEMP_DIR} as temporary directory"
   cp node_modules/feather-icons/dist/icons/* "$TEMP_DIR"
+
+  # Save XQuartz listen setting
+  XQUARTZ_SETTING=`defaults read org.macosforge.xquartz.X11.plist nolisten_tcp`
+  # Change XQuartz listen setting
+  defaults write org.macosforge.xquartz.X11.plist nolisten_tcp 0
+  # Start XQuartz server
+  xquartz &
+  # Save XQuartz pid
+  XQUARTZ_PID=${!}
+  export DISPLAY=0:0
+  # Disable X11 server authentication
+  xhost +

+  # Add --with-gui to Inkscape command
-  CMD="${PARALLEL} --bar ${INKSCAPE} -f {} --verb=EditSelectAll --verb=StrokeToPath --verb=FileSave --verb=FileQuit ::: ${TEMP_DIR}/*.svg"
+  CMD="${PARALLEL} --bar ${INKSCAPE} --with-gui -f {} --verb=EditSelectAll --verb=StrokeToPath --verb=FileSave --verb=FileQuit ::: ${TEMP_DIR}/*.svg"
   if [[ -x "${XVFB}" ]]; then ${XVFB} -d ${CMD}; else ${CMD}; fi

+  # Kill server
+  kill -SIGTERM ${XQUARTZ_PID}
+  # Reset XQuartz listen setting
+  defaults write org.macosforge.xquartz.X11.plist nolisten_tcp ${XQUARTZ_SETTING}
+
   print "Cleaning older installations..."
   rm -rf ${FEATHER_DIR} || true
   if [ -d "${FEATHER_DIR}" ]; then show_error "Can't remove Feather temp directory."; fi
```
Could check if running on macOS using `uname -s` and checking if result is `Darwin` or `Linux`.

The last problem I noticed is that the terminal prints the following:
```
\e[32mOK
\e[0mGenerating JS file mapping...\e[32mOK
\e[0mMoving TTF font to right path...\e[32mOK
\e[32mIf you see this message, everything is OK.
```
This is apparently due to Apple shipping macOS with outdated software, in this case Bash version 3.2 which does not support `\e` escape sequences... According to [this SO answer](https://stackoverflow.com/a/20983251) you can use `echo "$(tput setaf 2)OK"` to print green text in a cross-platform (unix) way.

rfbezerra commented:
Can you check in your installation if fil exists: _/usr/X11/bin/Xvfb_
In my OSx installation, it is default with installation of XQuartz.
If this is default location, I can use like xvfb-run to speed up and simplify process.
If yout confirm that, I do the changes on script.

hampustagerud commented:
It does exist, however homebrew installs it in `/opt/X11/bin/Xvfb` and I think `/usr/X11/bin/Xvfb` is the default installation included with macOS or something? But yes, it exists. I guess you could just use `which Xvfb` to obtain whatever path the user is using?

rfbezerra commented:
`which` way didnt find executable because these directories is not in the PATH.
I will check for a input environment variable _XVFB_, if not set, will try the `/usr/X11` path.
I will use the same approach to *nix and OsX

hampustagerud commented:
Sounds good 🙂

rfbezerra commented:
I updated script generator. Now, it should work in both, Linux and Mac OSx.
Changes:
- replaced `echo` with `printf`
- check for `XVFB` environment variable and use it if supplied

oblador commented:
Good to merge @hampustagerud?

hampustagerud commented:
Still getting a variety of errors on every glyph but I checked the resulting font in fontforge and the glyphs seemed fine so I'll merge it 🙂 Thanks @rfbezerra for your patience and work 👍
